Project sorting improvements

Rylern opened the pull request:
This PR is linked with issues #1537 and #1289 and implements [Alan's idea](https://github.com/qupath/qupath/issues/1537#issuecomment-2141451238).

It adds a [`Manager`](https://github.com/qupath/qupath/blob/7b272e5ee4c79d27dcf4e332d26ea4d0e5c80e3a/qupath-core/src/main/java/qupath/lib/projects/ResourceManager.java#L69) to the [`Project`](https://github.com/qupath/qupath/blob/main/qupath-core/src/main/java/qupath/lib/projects/Project.java) interface through the `Project.getSortingKeys()` function. The names of this manager correspond to sorting keys, while their values are string representations of booleans corresponding to whether the key is in ascending or descending order.

The current state of this PR makes project sorting persists when projects are closed / reopened.

It also enable to store several sorting keys and to define the sorting order (ascending or descending) for each of these keys, so that we could achieve this:
![image](https://github.com/user-attachments/assets/a7f6c570-dc2f-40f4-8d85-d88bdb392871)
 
However, the UI has to be changed to allow this. I can work on it, but I wanted to have your opinions on the current state of this PR first.

petebankhead commented:
The use of a manager still feels a bit strange to me, but I'm open to being convinced.

It's a reasonable mechanism if we want to ensure that sorting is something handled *entirely* by the UI, and isn't an inherent part of projects. But then we should use the existing [`Project.getResources(String, Class, String)`](https://qupath.github.io/javadoc/docs/qupath/lib/projects/Project.html#getResources(java.lang.String,java.lang.Class,java.lang.String)) rather than introduce a new `Project.getSortingKeys()` to the `Project` interface.

I think it's also preferable to store all such resources within a `resources` subdirectory, as done [here](https://github.com/qupath/qupath/blob/db71980322fbdcb47ad54599ef71d21d8f2c5f94/qupath-gui-fx/src/main/java/qupath/lib/display/settings/DisplaySettingUtils.java#L201). With this PR, I get a new `sorting_keys` directory directly within the project. I also got a file `SORT_KEY[IMAGE_NAME].txt` with the only content `true`; I was expecting e.g. a `sort_keys.txt` file with a list of the keys that are used.

The purpose of `ResourceManager` (for me at least) is that it supports having several named resources, typically implementations of a specific serializable class. So we might have something like this:

```java
class SortingKeys {
   List<String> keys;

   List<String> getKeys() {
      return Collections.unmodifiableList(keys);
   }

}

var sortingManager = project.getResources("resources/sorting_keys", SortingKeys.class, "json");
var sortKeys = sortingManager.contains("sortKeys") ? sortingManager.getResource("sortKeys).getKeys() : Collections.emtpyList();
```
but then it is still 'unconventional' to use a `Manager` when we only have a single resource with a fixed name (here, `"sortKeys"`).

---

What not use the alternative approach of extending the `Project` interface to implement [`MetadataStore`](https://github.com/qupath/qupath/blob/df21dcbaab953837d40aabd856d14b79faca6b7b/qupath-core/src/main/java/qupath/lib/objects/MetadataStore.java#L37) - or adding `String` metadata support similar to how it already works with [`ProjectImageEntry`](https://github.com/qupath/qupath/blob/df21dcbaab953837d40aabd856d14b79faca6b7b/qupath-core/src/main/java/qupath/lib/projects/ProjectImageEntry.java#L49)?

Then the UI can store keys as a list of strings easily, with any key of its choosing and the result serialized within the project.

To me, a `Map<String, String>` feels simpler and more intuitive. The advantages I see of a `Manager` are:
1. It avoids adding to the `project.qpproj` file size (in case someone attempts to add ridiculously-large strings)
2. The sort order wouldn't be lost if a project is saved in an older version of QuPath

But these are only an issue of the `Map` is serialized directly within `project.qpproj`. We *could* potentially overcome both of these by storing the `Map` in a separate `.json` file anyway within `resources/sorting` - without needing to rely on `Manager` to do so.

What do you think?

petebankhead commented on qupath-core/src/main/java/qupath/lib/projects/Project.java:
```diff
@@ -283,13 +284,29 @@ public default Manager<PixelClassifier> getPixelClassifiers() {
+	 * @param key the key associated with the value to retrieve
+	 * @return the value associated with the provided key, or an empty Optional if no such value exists
+	 */
+	Optional<String> getMetadataValue(String key);
```
I'm not sure `Optional` is required - the defined behavior for a `Map` is to return `null` if there is no value for the key.

The `Optional` doesn't resolve the ambiguity as to whether there is no value, or if the stored value is `null` (and there is no method to query `contains`).

petebankhead reviewed (changes requested):
I've added some comments.

One lingering question is to what extent this should be modelled on the [`Map` interface](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/Map.html), with similar behavior.

I wasn't entirely consistent myself, and we currently have:
* A [`MetadataStore`](https://github.com/qupath/qupath/blob/df21dcbaab953837d40aabd856d14b79faca6b7b/qupath-core/src/main/java/qupath/lib/objects/MetadataStore.java#L37), which is used by `PathObject` implementations
* Similar metadata support within [`ProjectImageEntry`](https://github.com/qupath/qupath/blob/df21dcbaab953837d40aabd856d14b79faca6b7b/qupath-core/src/main/java/qupath/lib/projects/ProjectImageEntry.java#L49)... but without implementing `MetadataStore`.

I think we should try to use this as an opportunity to improve consistency rather than have `Project` diverge further.

With that in mind, I think here we should do one of these:
* Use the `MetadataStore` approach, i.e. `Project extends MetadataStore`
* Use the `ProjectImageEntry` approach (without extending anything, but ensuring consistency of method names and signatures)
* Do *both* by updating `ProjectImageEntry` to extend `MetadataStore`
* Ignore the existing QuPath-specific strategies, and instead add a `Map<String, String> getMetadata()` method. This would return a 'proper' `Map` implementation that the caller can freely modify, and then rely on `project.syncChanges()` to make any metadata changes persistent.

The last option is quite convenient, since it provides useful `putAll` and `computeIfAbsent` methods - but may lead to `project.qpproj` being written more than it needs to be.

Rylern commented:
> Do both by updating ProjectImageEntry to extend MetadataStore

I think I'd prefer that option. The only problem I had with `MetadataStore` is that it doesn't have a method to delete a value. Can I add one?

petebankhead commented:
We can potentially be more ruthless than that... [`MetadataStore`](https://qupath.github.io/javadoc/docs/qupath/lib/objects/MetadataStore.html) currently supports non-String objects - but, weirdly, contains the method
```java
Map<String, String> getMetadataMap()
```
that assumes `String` only. The map is documented to be unmodifiable, but [at least once in QuPath's code it is assumed to be modifiable](https://github.com/qupath/qupath/blob/df21dcbaab953837d40aabd856d14b79faca6b7b/qupath-core/src/main/java/qupath/lib/objects/PathObjectTools.java#L1527).

*If* we don't need non-string objects (we'd need to check usages), I am thinking that we could deprecate all other methods in `MetadataStore`, leaving only `getMetadataMap()` to return a *modifiable* `Map<String, String>` - and provide default implementations for the other methods.

Then, in this case, you could potentially provide a `Map` implementation that takes care of JSON serialization by subclassing [`AbstractMap`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/Map.html#get(java.lang.Object)).

Since I was responsible for the original `MetadataStore` mess (and the associated [`MetadataMap`](https://github.com/qupath/qupath/blob/df21dcbaab953837d40aabd856d14b79faca6b7b/qupath-core/src/main/java/qupath/lib/objects/MetadataMap.java#L49)), I can look into how changing the interface might impact `PathObject`, where it is currently used.

*If* you think this is a sensible approach, then you could proceed here by adding
```java
public Map<String, String> getMetadataMap() // Or getMetadata() - see below
```
and handling its implementation and (JSON) serialization - then any common interface that includes that method can be added later to `Project`, `ProjectImageEntry` and `PathObject`.

What do you think?

One benefit of using a standard `Map` implementation is that Groovy scripting is nicer, e.g. see
* https://github.com/qupath/qupath/issues/1085

which is why we can now use `pathObject.measurements['Something'] = 2.0`.

In fact, if we simplified the name further to `getMetadata()` the this should work in Groovy:
```
pathObject.metadata['Something'] = 'A value'
project.metadata['Something'] = 'A value'
entry.metadata['Something'] = 'A value'
```

Rylern commented:
OK. I just have a question: if a class implementing `MetadataStore` returns a modifiable `Map<String, String>`, then the class won't know when the Map is modified. Isn't it a problem for situations where modifying the map should have side effects (such as saving the map to a file for example)?

Rylern commented:
I updated the `MetadataStore` and changed `Project` and `DefaultProject` accordingly.

I didn't update other classes that were using `MetadataStore` like `TMACoreObject`, and I didn't make `ProjectImageEntry` implements `MetadataStore`. Should I do that here, or create another PR?

petebankhead commented:
> Should I do that here, or create another PR?

No, I think this PR should focus on adding `Project.getMetadata()` - anything else can/should be done later.

Before going further with the general metadata support, I'd like to make sure that I was never using an `Object` with the `MetadataStore` within `PathObject`, in case that could be an issue. We also discussed having a new repo and module (`qupath-lib` perhaps?) similar to `qupath-fxtras` that contains generally-useful stuff that we can make permissively-licensed (i.e. Apache rather than GPL). I'd like to investigate adding a new, simple `MetadataStore` interface to that module, so that we can deprecate the old interface entirely - but I don't want to rush that, and need to find an hour or two to think about it properly.
